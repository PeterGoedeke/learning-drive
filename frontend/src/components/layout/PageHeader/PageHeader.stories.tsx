import { IconButton } from '@mui/material';
import { Meta, Story } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import PageHeader, { PageHeaderProps } from '.';

import SearchIcon from '../../icons/SearchIcon';

export default {
  component: PageHeader,
  title: 'Components/PageHeader',
  parameters: { layout: 'fullscreen' },
} as Meta;

interface PageHeaderStoryProps extends PageHeaderProps {}

const Template: Story<PageHeaderStoryProps> = (args) => (
  <BrowserRouter>
    <PageHeader {...args} />
  </BrowserRouter>
);

export const Default = Template.bind({});
Default.args = {
  title: 'Global Feed',
};

export const WithAction = Template.bind({});
WithAction.args = {
  title: 'Global Feed',
  action: (
    <IconButton color='secondary'>
      <SearchIcon />
    </IconButton>
  ),
};

export const WithBackButton = Template.bind({});
WithBackButton.args = {
  title: 'Global Feed',
  backButton: true,
};
