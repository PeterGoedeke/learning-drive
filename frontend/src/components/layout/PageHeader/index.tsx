import {
  Box,
  Divider,
  IconButton,
  Stack,
  styled,
  Toolbar,
  Tooltip,
  Typography,
} from '@mui/material';
import { ReactNode } from 'react';
import { useNavigate } from 'react-router-dom';

import ArrowLeftIcon from '../../icons/ArrowLeftIcon';

export interface PageHeaderProps {
  title: string;
  action?: ReactNode;
  backButton?: boolean;
}

/**
 * Displays a header at the top of the page. This header will be sticky even as the user scrolls.
 * It also can optionally be passed an action which will be displayed on the right side, as well
 * as a back button.
 */
const PageHeader = ({ title, action, backButton = false }: PageHeaderProps) => {
  const navigate = useNavigate();
  return (
    <AppBar>
      <Stack component={Toolbar} direction='row' spacing={1}>
        {backButton && (
          <Tooltip title='Go Back'>
            <IconButton
              sx={{ color: 'white' }}
              onClick={() => navigate(-1)}
              data-testid='back-button'
            >
              <ArrowLeftIcon fontSize='inherit' />
            </IconButton>
          </Tooltip>
        )}
        <Typography variant='h5' component='h1'>
          {title}
        </Typography>
        <Box sx={{ flexGrow: 1 }} />
        {action}
      </Stack>
      <Divider />
    </AppBar>
  );
};

const AppBar = styled(Box)(({ theme }) => ({
  position: 'sticky',
  top: 0,
  background: theme.palette.background.default,
  zIndex: theme.zIndex.appBar,
}));

export default PageHeader;
