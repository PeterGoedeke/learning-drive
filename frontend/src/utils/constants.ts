// Layout constants
export const SIDEBAR_WIDTH = 240;
export const PAGE_MARGIN = 3;
export const LOGIN_BUTTON_WIDTH = 220;