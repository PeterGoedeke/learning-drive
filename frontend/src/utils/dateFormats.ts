export const dateFormats = {
  post: 'h:mm a d/MM/yyyy',
};
