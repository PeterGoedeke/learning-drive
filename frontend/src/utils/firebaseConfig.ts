export const firebaseConfig = {
  apiKey: 'AIzaSyDkfHM81Rqs8dcUCRZVvgV0abTtGmdac1I',
  authDomain: 'learning-drive-fe98e.firebaseapp.com',
  projectId: 'learning-drive-fe98e',
  storageBucket: 'learning-drive-fe98e.appspot.com',
  messagingSenderId: '287598871941',
  appId: '1:287598871941:web:aff78432cae60e3db530ec',
  measurementId: 'G-C2T226SSLD',
};
