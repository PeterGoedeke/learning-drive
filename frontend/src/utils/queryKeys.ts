export const queryKey = {
  CATEGORIES: 'categories',
  POSTS: (page: number): [string, number] => ['posts', page],
};
