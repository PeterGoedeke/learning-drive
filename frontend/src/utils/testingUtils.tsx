import { PropsWithChildren } from 'react';
import { MemoryRouter, Location } from 'react-router-dom';

interface composeTestingWrapperOpts {
  initialEntries?: (string | Partial<Location>)[];
  initialIndex?: number;
}

export const composeTestingWrapper = ({
  initialEntries,
  initialIndex,
}: composeTestingWrapperOpts = {}) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>) => (
    <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
      {children}
    </MemoryRouter>
  );
  return Wrapper;
};
