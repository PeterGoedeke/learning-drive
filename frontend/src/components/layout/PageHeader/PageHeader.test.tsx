import { render, screen } from '@testing-library/react';

import PageHeader from '.';

import { composeTestingWrapper } from '../../../utils/testingUtils';

describe('PageHeader', () => {
  it('renders without error', async () => {
    const view = render(<PageHeader title='Global Feed' />, { wrapper: composeTestingWrapper() });

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders the correct title', async () => {
    render(<PageHeader title='Global Feed' />, { wrapper: composeTestingWrapper() });

    expect(screen.getByText('Global Feed')).toBeInTheDocument();
  });

  it('renders a back button when specified', async () => {
    const view = render(<PageHeader title='Global Feed' backButton />, {
      wrapper: composeTestingWrapper(),
    });

    expect(screen.getByTestId('back-button')).toBeInTheDocument();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders the action passed', async () => {
    const action = <button data-testid='action'>Action</button>;
    const view = render(<PageHeader title='Global Feed' action={action} />, {
      wrapper: composeTestingWrapper(),
    });

    expect(screen.getByTestId('action')).toBeInTheDocument();

    expect(view.asFragment()).toMatchSnapshot();
  });
});
