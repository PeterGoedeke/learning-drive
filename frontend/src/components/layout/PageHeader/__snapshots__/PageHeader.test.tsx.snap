// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader renders a back button when specified 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-j2nflj"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1uiyf5x-MuiToolbar-root-MuiStack-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1jfmqnk-MuiButtonBase-root-MuiIconButton-root"
        data-testid="back-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="arrowLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4 12C4 11.4477 4.44772 11 5 11H19C19.5523 11 20 11.4477 20 12C20 12.5523 19.5523 13 19 13H5C4.44772 13 4 12.5523 4 12Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M12.7071 4.29289C13.0976 4.68342 13.0976 5.31658 12.7071 5.70711L6.41421 12L12.7071 18.2929C13.0976 18.6834 13.0976 19.3166 12.7071 19.7071C12.3166 20.0976 11.6834 20.0976 11.2929 19.7071L4.29289 12.7071C3.90237 12.3166 3.90237 11.6834 4.29289 11.2929L11.2929 4.29289C11.6834 3.90237 12.3166 3.90237 12.7071 4.29289Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Global Feed
      </h1>
      <div
        class="MuiBox-root css-i9gxme"
      />
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
  </div>
</DocumentFragment>
`;

exports[`PageHeader renders the action passed 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-j2nflj"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1uiyf5x-MuiToolbar-root-MuiStack-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Global Feed
      </h1>
      <div
        class="MuiBox-root css-i9gxme"
      />
      <button
        data-testid="action"
      >
        Action
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
  </div>
</DocumentFragment>
`;

exports[`PageHeader renders without error 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-j2nflj"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1uiyf5x-MuiToolbar-root-MuiStack-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Global Feed
      </h1>
      <div
        class="MuiBox-root css-i9gxme"
      />
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
  </div>
</DocumentFragment>
`;
